ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java

```diff
@@ -208,30 +208,42 @@ public void resourceChanged(IResourceChangeEvent event) {
 		IResourceDelta delta = event.getDelta();
 
 		try {
+			List<IFile> added = new ArrayList<>();
+			List<IFile> removed = new ArrayList<>();
 			delta.accept(delta1 -> {
-				int kind = delta1.getKind();
 				IResource resource = delta1.getResource();
 				if (resource instanceof IFile) {
+					int kind = delta1.getKind();
 					if (kind == IResourceDelta.ADDED)
-						asyncStructureChanged((IFile) resource, true);
+						added.add((IFile) resource);
 					else if (kind == IResourceDelta.REMOVED)
-						asyncStructureChanged((IFile) resource, false);
+						removed.add((IFile) resource);
 					return false;
 				}
 				return true;
 			});
+			if (!added.isEmpty() || !removed.isEmpty()) {
+				asyncStructureChanged(added, removed);
+			}
 		} catch (CoreException e) {
 			PDEPlugin.logException(e);
 		}
 	}
 
-	private void asyncStructureChanged(final IFile file, final boolean added) {
+	private void asyncStructureChanged(List<IFile> added, List<IFile> removed) {
 		if (editor == null || editor.getEditorSite() == null)
 			return;
 		Shell shell = editor.getEditorSite().getShell();
 		Display display = shell != null ? shell.getDisplay() : Display.getDefault();
 
-		display.asyncExec(() -> structureChanged(file, added));
+		display.asyncExec(() -> {
```

Comment:
I would propose here is to schedule UIJob, which shows progress, checks for cancellation and so can be terminated by user in case the loop never ends here (100000 changes may take a while to process - and we are on UI thread).

Reply:
Well, the bug says: there is a huge number of UI tasks, and so moving from 100_000 single tasks that block because of the inefficient scheduling to a single task that blocks because of the work done by editor is not a big improvement.

So if we would give user a chance to cancel, so why not? However, I've not tried, may be not a UI job but ProgressService would be better here - because cancellation of a job if UI is busy is surely not possible. 

Something like `PlatformUI.getWorkbench().getProgressService().runInUI(PDEPlugin.getActiveWorkbenchWindow(), runnable, null);`

Ideally I would check if the `structureChanged()` need to run in UI at all, or to which extent it depends on UI and move as much as possible work from UI thread.

Reply:
How much baggage do I need for a UIJob?  E.g., can it be like this:

https://github.com/eclipse-pde/eclipse.pde/blob/96d307ba51b304cf51bd245e5f00a08657050c6c/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/wizards/target/TargetDefinitionContentPage.java#L271

Or would I need a new job name that's translatable?

And again, what's the effect of cancel on the state displayed to the user?